### PR TITLE
FAI-12009 | GitHub auth app params cleanup

### PR DESCRIPTION
**Pull request by dbruno21**
## Description

- Removed installation_id from app auth params
  - Replaced with multiple octokit instances (one per installation, which makes sense since quota is per installation as well)
- Removed client_id + client_secret from app auth params
  - Not used

## Type of change
- [ ] Bug fix
- [X] New feature
- [ ] Breaking change

### CODE REVIEW

File: sources/github-source/resources/spec.json [hunk 1]
```diff
@@ -174,25 +116,24 @@
         "description": "GitHub API previews. See https://docs.github.com/en/enterprise-server@3.1/rest/overview/api-previews",
         "items": {
           "type": "string"
-        },
-        "order": 4
+        }
       },
       "api_url": {
-        "order": 5,
+        "order": 6,
```
Comment (chalenge) [inline sources/github-source/resources/spec.json:122]: why do we need this in the source?

File: sources/github-source/src/github.ts [hunk 2]
```diff
@@ -25,32 +26,57 @@ export const DEFAULT_GRAPH = 'default';
 export const PAGE_SIZE = 100;
 
 export class GitHub {
-  private static github: GitHub;
+  private static _instance: GitHub;
```
Comment (chalenge) [inline sources/github-source/src/github.ts:29]: let's keep this as github for consistency with other sources

File: sources/github-source/src/github.ts [hunk 3]
```diff
@@ -25,32 +26,57 @@ export const DEFAULT_GRAPH = 'default';
...
   ) {}
 
   static async instance(
     cfg: GitHubConfig,
     logger: AirbyteLogger
   ): Promise<GitHub> {
-    if (GitHub.github) return GitHub.github;
-
-    const octokit = makeOctokitClient(cfg, logger);
+    let instance: GitHub = GitHub._instance;
```
Comment (chalenge) [inline sources/github-source/src/github.ts:42]: let's invert this and avoid lots of code in if statement and exit early

File: sources/github-source/src/github.ts [hunk 4]
```diff
@@ -64,9 +90,9 @@ export class GitHub {
   }
 
   async *getOrganizationsIterator(): AsyncGenerator<Organization> {
-    if (this.authType === 'token') {
-      const iter = this.octokit.paginate.iterator(
-        this.octokit.orgs.listForAuthenticatedUser,
+    if (this.config.authentication.type === 'token') {
```
Comment (chalenge) [inline sources/github-source/src/github.ts:93]: really not loving this, not too keen on us having to constantly check the config. Think we can do some OOP for `apps` and one for `orgs` and have these methods specific to that and avoid having if statements?

File: sources/github-source/src/github.ts [hunk 5]
```diff
@@ -25,32 +26,57 @@ export const DEFAULT_GRAPH = 'default';
...
-    if (this.authType === 'token') {
-      await this.octokit.users.getAuthenticated();
-    } else if (this.authType === 'app') {
-      await this.octokit.apps.getAuthenticated();
+    if (this.config.authentication.type === 'token') {
+      await this.authOctokit.users.getAuthenticated();
+    } else if (this.config.authentication.type === 'app') {
+      await this.authOctokit.apps.getAuthenticated();
+    } else {
+      throw new VError('Invalid authentication');
```
Comment (chalenge) [inline sources/github-source/src/github.ts:79]: `Unsupported authentication method: `${xxx}`

File: sources/github-source/src/github.ts [hunk 6]
```diff
@@ -25,32 +26,57 @@ export const DEFAULT_GRAPH = 'default';
...
+    return instance;
+  }
 
-    GitHub.github = new GitHub(octokit, cfg.authentication.auth, logger);
-    await GitHub.github.checkConnection();
-    return GitHub.github;
+  private async octokit(org?: string): Promise<Octokit> {
+    if (this.config.authentication.type !== 'app' || !org)
+      return this.authOctokit;
+    if (!this.octokitByInstallationOrg.has(org)) {
```
Comment (chalenge) [inline sources/github-source/src/github.ts:55]: given we already know the org list, feel like we should generate these Octokits when we instantiate `GitHub` so we don't have to do create [everytime we need it like](https://github.com/faros-ai/airbyte-connectors/pull/1543/files#diff-c3ec475eeb0ddb699ccf89b07eb40f24741bc8974d505f77035e1a8d27f4e563R129), not sure what other places we gonna need that too

File: sources/github-source/src/github.ts [hunk 7]
```diff
@@ -25,32 +26,57 @@ export const DEFAULT_GRAPH = 'default';
...
+      GitHub._instance = instance;
+    }
+    return instance;
+  }
 
-    GitHub.github = new GitHub(octokit, cfg.authentication.auth, logger);
-    await GitHub.github.checkConnection();
-    return GitHub.github;
+  private async octokit(org?: string): Promise<Octokit> {
+    if (this.config.authentication.type !== 'app' || !org)
```
Comment (chalenge) [inline sources/github-source/src/github.ts:53]: shouldn't this just be `if (this.config.authentication.type === 'token')`

File: sources/github-source/resources/spec.json (hunk 1 above)
Comment (dbruno21) [reply sources/github-source/resources/spec.json:122]: For the copilot seats logic to set inactive to true when the API doesn't return anymore the seat
https://github.com/faros-ai/airbyte-connectors/blob/a2f407c96a6b96edc336177e25214dee1d991640/sources/github-source/src/github.ts#L139-L165

Would it be doable in the converter instead?

Comment (dbruno21) [reply sources/github-source/resources/spec.json:122]: I talked to @ypc-faros about this and will try to move to destination on another PR

File: sources/github-source/src/github.ts (hunk 5 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts]: removed this piece of code (handled by oop)

File: sources/github-source/src/github.ts (hunk 6 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts]: updated to generate Octokits on GitHub instantiation

File: sources/github-source/src/github.ts (hunk 7 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts]: removed optional from org param

File: sources/github-source/src/github.ts (hunk 4 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts]: PTAL

File: sources/github-source/src/github.ts [hunk 8]
```diff
@@ -25,85 +26,53 @@ export const DEFAULT_GRAPH = 'default';
...
-      await this.octokit.apps.getAuthenticated();
-    }
-  }
+  abstract getAccessibleOrganizationsIterator(): AsyncGenerator<string>;
 
   @Memoize()
-  async getOrganizations(): Promise<ReadonlyArray<Organization>> {
-    const orgs: Organization[] = [];
-    for await (const org of this.getOrganizationsIterator()) {
+  async getAccessibleOrganizations(): Promise<ReadonlyArray<string>> {
```
Comment (chalenge) [inline sources/github-source/src/github.ts]: let's keep as `getOrganizations`

File: sources/github-source/src/github.ts [hunk 9]
```diff
@@ -194,3 +163,97 @@ export class GitHub {
     }
   }
 }
+
+export class TokenGitHub extends GitHub {
```
Comment (chalenge) [inline sources/github-source/src/github.ts:167]: maybe better names are GitHubTokenClient and GitHubAppClient. 
also nit: no need to export these

File: sources/github-source/src/github.ts [hunk 10]
```diff
@@ -25,85 +26,53 @@ export const DEFAULT_GRAPH = 'default';
...
+  }
 
-    const octokit = makeOctokitClient(cfg, logger);
+  abstract checkConnection(): Promise<void>;
 
-    GitHub.github = new GitHub(octokit, cfg.authentication.auth, logger);
-    await GitHub.github.checkConnection();
-    return GitHub.github;
-  }
+  abstract octokit(org: string): Octokit;
```
Comment (chalenge) [inline sources/github-source/src/github.ts:55]: this can be `octokit(org?: string)`

File: sources/github-source/src/github.ts [hunk 11]
```diff
@@ -194,3 +163,97 @@ export class GitHub {
...
+    cfg: GitHubConfig,
+    logger: AirbyteLogger
+  ): Promise<GitHub> {
+    const baseOctokit = makeOctokitClient(cfg, undefined, logger);
+    const github = new TokenGitHub(cfg, baseOctokit, logger);
+    await github.checkConnection();
+    return github;
+  }
+
+  octokit(org: string): Octokit {
```
Comment (chalenge) [inline sources/github-source/src/github.ts:178]: `octokit()` - our linters suck. It should be complaining

**Review by chalenge — APPROVED** [review]
Few things to address, but lgtm

File: sources/github-source/src/github.ts (hunk 10 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts:55]: I thought about it and think it's better to force org to be passed for developers to be conscious about it. I think there is no case where we would like to access resources outside of an installation, but if that happens it's better to access baseOctokit directly

File: sources/github-source/src/github.ts (hunk 9 above)
Comment (dbruno21) [reply sources/github-source/src/github.ts]: Had to export for mocking stuff in tests

**Pull request merged**